File: tree/l102.ts
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val === undefined ? 0 : val)
        this.left = (left === undefined ? null : left)
        this.right = (right === undefined ? null : right)
    }
}


function levelOrder(root: TreeNode | null): number[][] {
    let res_node = new Array<Array<TreeNode>>();
    let queue = new Array<TreeNode>();
    let level = 0;
    if (root !== null) queue.push(root);
    while (queue.length !== 0) {
        res_node[level] = queue;
        let new_queue = [];
        res_node[level].forEach(v => {
            v.left !== null ? new_queue.push(v.left) : '';
            v.right !== null ? new_queue.push(v.right) : '';
        })
        queue = new_queue;
        level++;
    }
    return res_node.map(v => {
        let arr = [];
        for (let node of v) {
            arr.push(node.val);
        }
        return arr
    });
};

File: src/l102.cpp
#include "leetcode.hpp"
#include <queue>
#include <utility>
#include <vector>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    std::vector<std::vector<int>> levelOrder(TreeNode* root) {
        std::queue<std::pair<int, TreeNode*>> node_q;
        std::vector<std::vector<int>> res;
        node_q.push({1, root});
        while (!node_q.empty()) {
            auto [level, node] = node_q.front();
            node_q.pop();
            if (node == nullptr) continue;
            if (res.size() < level) res.push_back(std::vector<int>{});
            res[level - 1].push_back(node->val);
            node_q.push({level+1, node->left});
            node_q.push({level+1, node->right}); 
        }
        return res;
    }
};